### ISSUE TITLE
ERROR: UndefVarError: TH2_5 not defined when reading TH2F

### ISSUE THREAD

**Issue by Yuan-Ru-Lin** [JuliaHEP/UnROOT.jl#216]
Executing the test (originally posted by @aminnj in https://github.com/JuliaHEP/UnROOT.jl/issues/47#issuecomment-881813323_) gives the error message:

```julia
julia> x = UnROOT.TH2F(f.fobj, f.directory.keys[1], f.streamers.refs)
ERROR: UndefVarError: TH2_5 not defined
Stacktrace:
 [1] stream!(io::IOBuffer, fields::Dict{Symbol, Any}, ::Type{UnROOT.TH2}; check::Bool)
   @ UnROOT ~/.julia/packages/UnROOT/T6Krf/src/streamers.jl:580
 [2] TH(io::UnROOT.MmapStream, tkey::UnROOT.TKey32, refs::Dict{Int32, Any})
   @ UnROOT ~/.julia/packages/UnROOT/T6Krf/src/bootstrap.jl:813
 [3] TH2F(io::UnROOT.MmapStream, tkey::UnROOT.TKey32, refs::Dict{Int32, Any})
   @ UnROOT ~/.julia/packages/UnROOT/T6Krf/src/bootstrap.jl:801
 [4] top-level scope
   @ REPL[21]:1
```

The sample was generated by

```python
import ROOT as r
f = r.TFile("histograms.root", "recreate")
th2f = r.TH2F("myTH2F", "", 10, -2, 2, 10, -2, 2)
for x,y,w in [
        [-1.5, -1.5, 20.0],
        [+1.5, +1.5, 1.0],
        [-1.5, +1.5, 20.0],
        [+1.5, -1.5, 1.0],
        ]:
    th2f.Fill(x, y, w)
f.Write()
f.Close()
```

---

### UnROOT

v0.9.1

### Julia

```
Julia Version 1.8.3
Commit 0434deb161e (2022-11-14 20:14 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 80 × Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, cascadelake)
  Threads: 1 on 80 virtual cores
Environment:
  LD_LIBRARY_PATH = /cvmfs/belle.cern.ch/el7/externals/v01-11-01/Linux_x86_64/common/root/lib:/cvmfs/belle.cern.ch/el7/externals/v01-11-01/Linux_x86_64/common/lib64:/cvmfs/belle.cern.ch/el7/externals/v01-11-01/Linux_x86_64/opt/lib64:/cvmfs/belle.cern.ch/el7/externals/v01-11-01/Linux_x86_64/opt/lib:/cvmfs/belle.cern.ch/el7/externals/v01-11-01/Linux_x86_64/common/epics/lib/linux-x86_64:/cvmfs/belle.cern.ch/el7/externals/v01-11-01/Linux_x86_64/common/lib64:/cvmfs/belle.cern.ch/el7/externals/v01-11-01/Linux_x86_64/common/lib:/gpfs/home/belle/leonlin/B2Kstargamma/lib/Linux_x86_64/opt:/cvmfs/belle.cern.ch/el7/releases/light-2210-devonrex/lib/Linux_x86_64/opt:/sw/belle/local/neurobayes/lib/:/sw/belle/local/neurobayes/lib:/opt/lsf/10.1/linux3.10-glibc2.17-x86_64/lib:/cvmfs/belle.cern.ch/el7/externals/v01-11-01/Linux_x86_64/common/lib
  DYLD_LIBRARY_PATH = /cvmfs/belle.cern.ch/el7/externals/v01-11-01/Linux_x86_64/common/root/lib
```

### ROOT

```
   ------------------------------------------------------------------
  | Welcome to ROOT 6.24/06                        https://root.cern |
  | (c) 1995-2021, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for linuxx8664gcc on Jul 28 2022, 19:53:00                 |
  | From tag , 1 September 2021                                      |
  | With g++ (GCC) 11.2.0                                            |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q'       |
   ------------------------------------------------------------------
```

### Python3

Python 3.8.8

**Comment by Yuan-Ru-Lin** [comment]
`TH1F` works well, though.

I generated a TH1F in a similar fashion i.e.

```python
...
th1f = r.TH1F("myTH1F", "", 10, -2, 2)
for x,w in [
        [-1.5, 20.0],
        [+1.5, 1.0],
        [-1.5, 20.0],
        [+1.5, 1.0],
        ]:
    th1f.Fill(x, w)
...
```

and then I got

```julia
julia> UnROOT.TH1F(f.fobj, f.directory.keys[2], f.streamers.refs)
Dict{Symbol, Any} with 102 entries:
  :fZaxis_fTitleSize   => 0.065
  :fYaxis_fXmin        => 0.0
  :fYaxis_fNbins       => 1
...
```

**Comment by tamasgal** [comment]
Hmm, can you try https://github.com/JuliaHEP/UnROOT.jl/tree/fix-th2_5 ?

**Comment by Moelf** [comment]
in general whenever ROOT bumps the version in this case 4 to 5, they could contain layout or length change, so it's a matter of reading the difference in ROOT source code and see if it's a trivial change to add

**Comment by tamasgal** [comment]
Yes exactly. If it works with the "dirty fix" above, we might get away with it ;) but a thorough check is needed.

**Comment by tamasgal** [comment]
I think the "dirty fix" should be fine.

Here is where the class definition was changed from 4 to 5: https://github.com/root-project/root/commit/14ee26b7659b10c1e34f970e575cd3f1ec6084c8

it reads:

> At the moment, the derived class StreamerInfo contains the base class version number :(
This means we need to update the derived class version number when the base class number increases ...

and here is a related discussion: https://root-forum.cern.ch/t/problem-in-opening-past-rootfile-tclonesarray-with-th1s-with-root-v6-16-and-ubuntu18/33293

I am not sure if I understood correctly but it seems that the streamer logic is untouched.

**Comment by tamasgal** [comment]
I tried and it does something 😉 

```
julia> using UnROOT

julia> f = ROOTFile("test/samples/TH2_5.root")
ROOTFile with 1 entry and 15 streamers.
test/samples/TH2_5.root
└─ myTH2F (TH2F)


julia> h = UnROOT.TH2F(f.fobj, f.directory.keys[1], f.streamers.refs)
Dict{Symbol, Any} with 106 entries:
  :fZaxis_fTitleSize   => 0.035
  :fYaxis_fXmin        => -2.0
  :fYaxis_fNbins       => 10
  :fZaxis_fLast        => 0
  :fYaxis_fName        => "yaxis"
  :fMaximum            => -1111.0
  :fZaxis_fLabelFont   => 42
  :fZaxis_fLabelSize   => 0.035
  :fZaxis_fXmin        => 0.0
  ...
  ...
```

**Comment by Yuan-Ru-Lin** [comment]
```julia
(test-unroot) pkg> add https://github.com/JuliaHEP/UnROOT.jl#fix-th2_5
julia> using UnROOT
julia> f = ROOTFile("histograms.root")
julia> x = UnROOT.TH2F(f.fobj, f.directory.keys[1], f.streamers.refs)
ERROR: MethodError: no method matching readfields!(::IOBuffer, ::Dict{Symbol, Any}, ::Type{UnROOT.TH2_5})
Closest candidates are:
  readfields!(::Any, ::Any, ::Type{UnROOT.TNamed_1}) at ~/.julia/packages/UnROOT/GKT9E/src/bootstrap.jl:6
  readfields!(::Any, ::Any, ::Type{UnROOT.TAttLine_1}) at ~/.julia/packages/UnROOT/GKT9E/src/bootstrap.jl:14
  readfields!(::Any, ::Any, ::Type{UnROOT.TAttLine_2}) at ~/.julia/packages/UnROOT/GKT9E/src/bootstrap.jl:20
  ...
Stacktrace:
 [1] stream!(io::IOBuffer, fields::Dict{Symbol, Any}, ::Type{UnROOT.TH2}; check::Bool)
   @ UnROOT ~/.julia/packages/UnROOT/GKT9E/src/streamers.jl:582
 [2] TH(io::UnROOT.MmapStream, tkey::UnROOT.TKey32, refs::Dict{Int32, Any})
   @ UnROOT ~/.julia/packages/UnROOT/GKT9E/src/bootstrap.jl:814
 [3] TH2F(io::UnROOT.MmapStream, tkey::UnROOT.TKey32, refs::Dict{Int32, Any})
   @ UnROOT ~/.julia/packages/UnROOT/GKT9E/src/bootstrap.jl:802
 [4] top-level scope
   @ REPL[7]:1
```

**Comment by tamasgal** [comment]
Sorry @Yuan-Ru-Lin I forgot to add the second commit, try again please with the latest branch!

**Comment by Yuan-Ru-Lin** [comment]
Yup it works now! That was quick!

**Comment by tamasgal** [comment]
Maybe you can add some value checks to the test suite, or just post them here and I can add them 😉

**Comment by Yuan-Ru-Lin** [comment]
Hmmm, it seems the author of the original comment has done that.

https://github.com/JuliaHEP/UnROOT.jl/blob/0ee44c1dc6058ba779fe9317d4b3f1a94955c701/test/runtests.jl#L554-L573

**Comment by tamasgal** [comment]
Yes, but that's for another file which uses `TH2_4` ;)

**Comment by Yuan-Ru-Lin** [comment]
Okay, let me borrow some of them to TH2_5.

**Comment by Yuan-Ru-Lin** [comment]
Done. I've reduced the size of both bins to make the test easier to write.

**Comment by tamasgal** [comment]
@all-contributors please add @Yuan-Ru-Lin for tests

**Comment by Yuan-Ru-Lin** [comment]
Let me pull the branch in my fork real quick

**Comment by tamasgal** [comment]
Just noticed that all-contributes stopped working when transferring UnROOT.jl to `JuliaHEP` 🙈 let me try again

**Comment by tamasgal** [comment]
@all-contributors please add @Yuan-Ru-Lin for tests

**Comment by Yuan-Ru-Lin** [comment]
Check out the PR https://github.com/JuliaHEP/UnROOT.jl/pull/219

**Comment by tamasgal** [comment]
Thanks @Yuan-Ru-Lin, all set :)
